particles/NonEqParticle.py

```diff
@@ -12,16 +12,141 @@
+#         return numpy.vectorize(lambda p: (
+#             particle.distribution(p) * p**4 / particle.conformal_energy(p)
+#             * particle.dof / 6. / numpy.pi**2 / particle.params.a**4
+#         ), otypes=[numpy.float_])
+#     return press()
+
+
+# def entropy(particle):
+#     @lambda_integrate(particle.grid.BOUNDS)
+#     def ent():
+#         """ ## Entropy
+
+#             \begin{equation}
+#                 s = - \int_0^\inf p^2 dp \left{ f(p) \ln f(p) \mp (1 \pm f(p)) \ln (1 \pm f(p)) \right}
+#             \end{equation}
+#         """
+
+#         def integrand(p):
+#             f = particle.distribution(p)
+#             eta = particle.eta
+
+#             if f == 0:
+#                 return 0.
+
+#             return (- particle.dof / 2 / numpy.pi**2 / particle.params.a**3
+#                     * p**2 * (f * numpy.log(f) + eta * (1 - eta * f) * numpy.log(1 - eta * f)))
+
+#         return numpy.vectorize(integrand, otypes=[numpy.float_])
+#     return ent()
+
+
+# # @lambda_integrate()
+# # def inverse_gamma_factor(particle):
+# #     """ ## Average gamma factor of the distribution """
+
+# #     def integrand(p):
+# #         if particle.mass == 0:
+# #             return 0.
+
+# #         return (particle.dof / 2 / numpy.pi**2 * p**2 / particle.params.a**3
+# #                 * particle.distribution(p) * particle.conformal_mass / particle.conformal_energy(p))
+
+# #     return numpy.vectorize(integrand, otypes=[numpy.float_])
+
+
+# """ ## Master equation terms """
+
+# """ ### Numerator
+
+#     \begin{equation}
+#         -\frac{g}{2 \pi^2} \int dy \, y^2 \sqrt{y^2 + \frac{M_N^2 x^2}{m^2}} I_{coll}(y)
+#     \end{equation}
+# """
+
+
+# def numerator(particle):
+#     if environment.get('SIMPSONS_INTEGRATION'):
+#         y = particle.grid.TEMPLATE
+#         return simps((
+#             -1. * particle.dof / 2. / numpy.pi**2
+#             * y**2 * particle.conformal_energy(y)
+#             * particle.collision_integral / particle.params.x
+#         ), y)
+#     else:
+#         integral = linear_interpolation(particle.collision_integral / particle.params.x,
+#                                         particle.grid.TEMPLATE)
+#         return lambda_integrate(particle.grid.BOUNDS)(lambda particle: numpy.vectorize(lambda y: (
+#             -1. * particle.dof / 2. / numpy.pi**2
+#             * y**2 * particle.conformal_energy(y)
+#             * integral(y)
+#         ), otypes=[numpy.float_]))(particle)
+
+
+# def denominator(particle):
+#     """
+#     ### Denominator
+
+#     \begin{equation}
+#         0
+#     \end{equation}
+#     """
+#     return 0.
+
+
+##################################### SIMPSONS INTEGRATION #####################################
 
-@lambda_integrate()
 def density(particle):
-    return numpy.vectorize(lambda y: (
-        particle.distribution(y) * y**2
-        * particle.dof / 2. / numpy.pi**2 / particle.params.a**3
-    ), otypes=[numpy.float_])
+    temp = particle.grid.TEMPLATE
```

Comment:
I think instead of rewriting all the methods you could change the internals of `lambda_integrate`. Also adding a switch for this might be a good idea as well. I'm okay with the default value being "simpson's" now

Reply:
I thought it would be nice to keep lambda_integrate as an 'independent' function. That it will only accept bounds. If you have something in mind, feel free to improve it.